### PR TITLE
implement optimism validator contracts

### PR DESCRIPTION
**Pull request by cds95**
**Description**
 The purpose of this change is to implement a way for Chainlink to update the status of the Optimism sequencer on L2.  This is done by calling the `validate` function in the `OptimismValidator` contract on L1, which will send a cross chain message to the `OptimismSequencerUptimeFeed` contract on L2 if there is a status change.
 
**High Level Changes**
 - Add `OptimismValidator` on L1 
 - Add `OptimismSequencerUptimeFeed` contract on L2 (Optimism)

**Testing**
 - Deploy contracts on testnet and verify that calling the `validate` function with an updated status relays the message onto L2.
 - Unit tests

### CODE REVIEW

File: contracts/src/v0.8/dev/OptimismSequencerUptimeFeed.sol [hunk 1]
```diff
@@ -0,0 +1,331 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import {AddressAliasHelper} from "./vendor/arb-bridge-eth/v0.8.0-custom/contracts/libraries/AddressAliasHelper.sol";
```
Comment (yosriady) [inline contracts/src/v0.8/dev/OptimismSequencerUptimeFeed.sol:4]: Please remove all Arbitrum dependencies

File: contracts/src/v0.8/dev/OptimismSequencerUptimeFeed.sol [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+    }
+  }
+
+  /**
+   * @notice Messages sent by the stored L1 sender will arrive on L2 with this
+   *  address as the `msg.sender`
+   * @return L2-aliased form of the L1 sender address
+   */
+  function aliasedL1MessageSender() public view returns (address) {
+    return AddressAliasHelper.applyL1ToL2Alias(l1Sender());
```
Comment (yosriady) [inline contracts/src/v0.8/dev/OptimismSequencerUptimeFeed.sol:163]: This helper is Arbitrum-specific. Is this function still relevant for Optimism?

File: contracts/src/v0.8/dev/OptimismSequencerUptimeFeed.sol [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+   *
+   * @param status Sequencer status
+   * @param timestamp Block timestamp of status update
+   */
+  function updateStatus(bool status, uint64 timestamp) external override {
+    FeedState memory feedState = s_feedState;
+    requireInitialized(feedState.latestRoundId);
+    if (
+      msg.sender != address(s_l2CrossDomainMessenger) &&
+      s_l2CrossDomainMessenger.xDomainMessageSender() != aliasedL1MessageSender()
```
Comment (yosriady) [inline contracts/src/v0.8/dev/OptimismSequencerUptimeFeed.sol:220]: The `aliasedL1MessageSender` was using Arbitrum's aliasing helper, is the current Optimism compatible with this?

File: contracts/src/v0.8/dev/OptimismValidator.sol [hunk 4]
```diff
@@ -1,71 +1,185 @@
...
-    require(l2FlagsAddr != address(0), "Invalid L2 Flags address");
-    CROSS_DOMAIN_MESSENGER = crossDomainMessengerAddr;
-    L2_CROSS_DOMAIN_FORWARDER = l2CrossDomainForwarderAddr;
-    L2_FLAGS = l2FlagsAddr;
+    require(l1CrossDomainMessengerAddress != address(0), "Invalid xDomain Messenger address");
+    require(l2UptimeFeedAddr != address(0), "Invalid OptimismSequencerUptimeFeed contract address");
+    L1_CROSS_DOMAIN_MESSENGER_ADDRESS = l1CrossDomainMessengerAddress;
+    L2_UPTIME_FEED_ADDR = l2UptimeFeedAddr;
+    // Additional L2 payment configuration
+    _setConfigAC(configACAddr);
```
Comment (yosriady) [inline contracts/src/v0.8/dev/OptimismValidator.sol:75]: Is this external access controller necessary? `setGasLimit` uses `onlyOwner` and not `onlyOwnerOrConfigAccess`

File: contracts/src/v0.8/dev/OptimismValidator.sol [hunk 5]
```diff
@@ -1,71 +1,185 @@
...
+  function setConfigAC(address accessController) external onlyOwner {
+    _setConfigAC(accessController);
   }
 
   /**
    * @notice validate method sends an xDomain L2 tx to update Flags contract, in case of change from `previousAnswer`.
-   * @dev A message is sent via the Optimism CrossDomainMessenger L1 contract. The "payment" for L2 execution happens on L1,
-   *   using the gas attached to this tx (some extra gas is burned by the Optimism bridge to avoid DoS attacks).
-   *   This method is accessed controlled.
+   * @dev A retryable ticket is created on the Optimism L1 Inbox contract. The tx gas fee can be paid from this
```
Comment (yosriady) [inline contracts/src/v0.8/dev/OptimismValidator.sol:180]: Does the optimism bridge refer to cross-chain messages as retryable tickets? We should use the terms they use.

File: contracts/src/v0.8/dev/OptimismValidator.sol [hunk 6]
```diff
@@ -1,71 +1,185 @@
...
+  }
+
+  /**
+   * @notice withdraws funds from L2 xDomain alias address (representing this L1 contract)
+   * @dev only owner can call this
+   * @param amount of funds to withdraws
+   * @param refundAddr address where gas excess on L2 will be sent
+   *   WARNING: `refundAddr` is not aliased! Make sure you can recover the refunded funds on L2.
+   */
+  function withdrawFundsFromL2(uint256 amount, address refundAddr) external onlyOwner {
```
Comment (yosriady) [inline contracts/src/v0.8/dev/OptimismValidator.sol:155]: Did you manage to test this flow on the testnet?

**Review by yosriady — CHANGES_REQUESTED** [review]

File: contracts/test/v0.8/dev/OptimismValidator.test.ts [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+        optimismValidator.connect(eoaValidator).validate(0, 0, 1, 1),
+      ).to.be.revertedWith('No access')
+    })
+
+    it('post sequencer offline', async () => {
+      await optimismValidator.addAccess(eoaValidator.address)
+
+      const now = Math.ceil(Date.now() / 1000) + 1000
+      await ethers.provider.send('evm_setNextBlockTimestamp', [now])
+      const arbitrumSequencerStatusRecorderCallData =
```
Comment (yosriady) [inline contracts/test/v0.8/dev/OptimismValidator.test.ts:162]:
```suggestion
      const sequencerStatusRecorderCallData =
```

File: contracts/test/v0.8/dev/OptimismValidator.test.ts [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+          true,
+          now,
+        ])
+
+      await expect(optimismValidator.connect(eoaValidator).validate(0, 0, 1, 1))
+        .to.emit(mockOptimismL1CrossDomainMessenger, 'SentMessage')
+        .withArgs(
+          L2_SEQ_STATUS_RECORDER_ADDRESS,
+          optimismValidator.address,
+          arbitrumSequencerStatusRecorderCallData,
```
Comment (yosriady) [inline contracts/test/v0.8/dev/OptimismValidator.test.ts:173]:
```suggestion
          sequencerStatusRecorderCallData,
```

**Review by yosriady — CHANGES_REQUESTED** [review]

File: contracts/src/v0.8/dev/OptimismValidator.sol [hunk 9]
```diff
@@ -1,71 +1,122 @@
...
+import "../interfaces/AggregatorV3Interface.sol";
 import "../SimpleWriteAccessController.sol";
 
 /* ./dev dependencies - to be moved from ./dev after audit */
 import "./interfaces/FlagsInterface.sol";
 import "./interfaces/ForwarderInterface.sol";
-import "./vendor/@eth-optimism/contracts/0.4.7/contracts/optimistic-ethereum/iOVM/bridge/messaging/iOVM_CrossDomainMessenger.sol";
+import "./interfaces/OptimismSequencerUptimeFeedInterface.sol";
+import "@eth-optimism/contracts/L1/messaging/IL1CrossDomainMessenger.sol";
+import "@eth-optimism/contracts/L2/messaging/IL2ERC20Bridge.sol";
```
Comment (krebernisak) [inline contracts/src/v0.8/dev/OptimismValidator.sol:15]: Is this import used?

Comment (cds95) [reply contracts/src/v0.8/dev/OptimismValidator.sol:15]: good catch no it is not

File: contracts/src/v0.8/dev/OptimismSequencerUptimeFeed.sol [hunk 10]
```diff
@@ -133,15 +137,16 @@ contract OptimismSequencerUptimeFeed is
...
+   * @param timestamp The L1 block timestamp of status update
    */
   function recordRound(
     uint80 roundId,
     bool status,
     uint64 timestamp
   ) private {
-    Round memory nextRound = Round(status, timestamp);
-    FeedState memory feedState = FeedState(roundId, status, timestamp);
+    uint64 recordedAt = uint64(block.timestamp);
```
Comment (yosriady) [inline contracts/src/v0.8/dev/OptimismSequencerUptimeFeed.sol:147]: Is there a reason why we use the term `recordedAt` for the updatedAt field? I wonder if introducing new terms like this is more confusing

Comment (cds95) [reply contracts/src/v0.8/dev/OptimismSequencerUptimeFeed.sol]: just renamed to `updatedAt`

File: contracts/src/v0.8/dev/OptimismValidator.sol [hunk 11]
```diff
@@ -1,71 +1,121 @@
...
+    return s_gasLimit;
+  }
+
+  /**
+   * @notice makes this contract payable
+   * @dev receives funds:
+   *  - to use them (if configured) to pay for L2 execution on L1
+   *  - when withdrawing funds from L2 xDomain alias address (pay for L2 execution on L2)
+   */
+  receive() external payable {}
```
[the remark this answers is not in the document]
Comment (yosriady) [reply contracts/src/v0.8/dev/OptimismValidator.sol]: Why? Because the L2 gas fees are subsidized?

[the remark this answers is not in the document]
Comment (cds95) [reply contracts/src/v0.8/dev/OptimismValidator.sol]: My thinking is that we currently don’t send any gas when we send cross chain messages unlike the Arbitrum validator. We currently just set the required gasLimit parameter.

**Review by yosriady — DISMISSED** [review]

**Review by alexroan — APPROVED** [review]

**Comment by cds95** [comment]
Closing due to unsigned commits.  New PR: https://github.com/smartcontractkit/chainlink/pull/6885

**Pull request closed without merging**